crates/astria-sequencer/src/mempool.rs

```diff
@@ -114,6 +128,74 @@ impl std::hash::Hash for EnqueuedTransaction {
     }
 }
 
+const TX_TTL: i64 = 600; // 10 minutes 
+const INVALID_CACHE_SIZE: usize = 4096;
+const TX_INVALID_TTL: i64 = 60; // 60 seconds
+
+/// `InvalidCache` provides for signaling to `CometBFT` what transactions
+///  can be removed from the `CometBFT` mempool.
+///
+/// The cache places a TTL on the entries, allowing users
+/// to request their transaction to be processed again after the
+/// ttl has expired. This is useful for users who want to re-run a transaction
+/// that failed to execute or to re-add a transaction that has expired.
+#[derive(Clone)]
+pub(crate) struct InvalidCache {
+    cache: HashSet<[u8; 32]>,
+    time_added: HashMap<[u8; 32], i64>,
+    remove_queue: VecDeque<[u8; 32]>,
+    max_size: usize,
+    time_to_live: i64,
+    size: usize,
+}
+
+impl InvalidCache {
+    fn new(max_size: usize, time_to_live: i64) -> Self {
+        Self {
+            cache: HashSet::new(),
+            time_added: HashMap::new(),
+            remove_queue: VecDeque::with_capacity(max_size),
+            max_size,
+            time_to_live,
+            size: 0,
+        }
+    }
+
+    fn cached(&self, tx_hash: [u8; 32]) -> bool {
+        // the tx is known and entry hasn't expired
+        self.cache.contains(&tx_hash)
+            && (Time::now().unix_timestamp()
+                <= self.time_added[&tx_hash]
+                    .checked_add(self.time_to_live)
+                    .expect("overflowed ttl add"))
+    }
```

Comment:
it might be nicer to store `expiry_time` instead of `time_added` when a tx is inserted, that way you don't need to do the add + expect here each time

Reply:
or maybe have separate functions for `is_cached` and `is_expired`?